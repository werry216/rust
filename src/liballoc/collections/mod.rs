// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

//! Collection types.

#![stable(feature = "rust1", since = "1.0.0")]

pub mod binary_heap;
mod btree;
pub mod linked_list;
pub mod vec_deque;

#[stable(feature = "rust1", since = "1.0.0")]
pub mod btree_map {
    //! A map based on a B-Tree.
    #[stable(feature = "rust1", since = "1.0.0")]
    pub use super::btree::map::*;
}

#[stable(feature = "rust1", since = "1.0.0")]
pub mod btree_set {
    //! A set based on a B-Tree.
    #[stable(feature = "rust1", since = "1.0.0")]
    pub use super::btree::set::*;
}

#[stable(feature = "rust1", since = "1.0.0")]
#[doc(no_inline)]
pub use self::binary_heap::BinaryHeap;

#[stable(feature = "rust1", since = "1.0.0")]
#[doc(no_inline)]
pub use self::btree_map::BTreeMap;

#[stable(feature = "rust1", since = "1.0.0")]
#[doc(no_inline)]
pub use self::btree_set::BTreeSet;

#[stable(feature = "rust1", since = "1.0.0")]
#[doc(no_inline)]
pub use self::linked_list::LinkedList;

#[stable(feature = "rust1", since = "1.0.0")]
#[doc(no_inline)]
pub use self::vec_deque::VecDeque;

/// An intermediate trait for specialization of `Extend`.
#[doc(hidden)]
trait SpecExtend<I: IntoIterator> {
    /// Extends `self` with the contents of the given iterator.
    fn spec_extend(&mut self, iter: I);
}
