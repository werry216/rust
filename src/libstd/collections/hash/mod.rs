//! Unordered containers, implemented as hash-tables

mod bench;
mod table;
pub mod map;
pub mod set;
