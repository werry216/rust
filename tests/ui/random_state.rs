#![warn(clippy::random_state)]

use std::collections::hash_map::RandomState;
use std::collections::hash_map::{DefaultHasher, HashMap};
use std::hash::{BuildHasherDefault};

fn main() {
    // Should warn
    let mut map = HashMap::new();
    map.insert(3, 4);
    let mut map = HashMap::with_hasher(RandomState::new());
    map.insert(true, false);
    let _map: HashMap<_, _> = vec![(2, 3)].into_iter().collect();
    let _vec: Vec<HashMap<i32, i32>>;
    // Shouldn't warn
    let _map: HashMap<i32, i32, BuildHasherDefault<DefaultHasher>> = HashMap::default();
    let mut map = HashMap::with_hasher(BuildHasherDefault::<DefaultHasher>::default());
    map.insert("a", "b");
}
