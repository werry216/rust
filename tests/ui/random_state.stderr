error: usage of RandomState
  --> $DIR/random_state.rs:9:19
   |
LL |     let mut map = HashMap::new();
   |                   ^^^^^^^^^^^^
   |
   = note: `-D clippy::random-state` implied by `-D warnings`

error: usage of RandomState
  --> $DIR/random_state.rs:11:19
   |
LL |     let mut map = HashMap::with_hasher(RandomState::new());
   |                   ^^^^^^^^^^^^^^^^^^^^

error: usage of RandomState
  --> $DIR/random_state.rs:13:15
   |
LL |     let _map: HashMap<_, _> = vec![(2, 3)].into_iter().collect();
   |               ^^^^^^^^^^^^^

error: usage of RandomState
  --> $DIR/random_state.rs:14:19
   |
LL |     let _vec: Vec<HashMap<i32, i32>>;
   |                   ^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

